### PR TITLE
add python object to json auto conversion feature, test and doc

### PR DESCRIPTION
**Pull request by qingfengxia**
it is a resent PR for the close#16,  to sort out the modification history

### CODE REVIEW

**Comment by qingfengxia** [comment]
2/4 failed 
CI : windows  error:
 C:\projects\pybind11-json\include\pybind11_json/pybind11_json.hpp(15): fatal error C1083: Cannot open include file: 'nlohmann/json.hpp': No such file or directory


OSX: 

/Users/travis/build/pybind/pybind11_json/include/pybind11_json/pybind11_json.hpp:15:10: fatal error: 'nlohmann/json.hpp' file not found
#include "nlohmann/json.hpp"

I seems have not change anything near that line.

File: include/pybind11_json/pybind11_json.hpp [hunk 1]
```diff
@@ -129,4 +129,48 @@ namespace nlohmann
...
 
+namespace pybind11
+{
+    namespace detail
+    {
+        template <> struct type_caster<nlohmann::json>
+        {
+        public:
+            /**
+             * This macro establishes the name 'user_type' in
```
Comment (martinRenou) [inline include/pybind11_json/pybind11_json.hpp]: The comment is not in sync with the code, is it?
Shouldn't `user_type` be `nlohmann_json` here?

File: include/pybind11_json/pybind11_json.hpp [hunk 2]
```diff
@@ -129,4 +129,48 @@ namespace nlohmann
...
+                /* Extract PyObject from handle */
+                /* Now try to convert into a C++ type */
+                value = nlohmann::detail::to_json_impl(src);
+                if (value.is_null())
+                    return false;
+                return true;
+            }
+
+            /**
+             * Conversion part 2 (C++ -> Python): convert an inty instance into
```
Comment (martinRenou) [inline include/pybind11_json/pybind11_json.hpp]: I think `inty` comes from the PYBIND11 documentation here, you might want to replace it with `nlohmann::json`. 

I would be fine removing all the comments though :)

**Comment by martinRenou** [comment]
I'm not ignoring this PR. If you agree, I can commit to your branch and make the changes I requested (while keeping you as the author of the commit so you get credits for it, of course).

**Comment by qingfengxia** [comment]
Sorry, I have been too busy, that I ignored/had not read your previous email.

If you had already had some code written, you can send PR to me, then I will merge. otherwise, I can do it tonight.


Qingfeng Xia
Senior Research Software Engineer, Culham Centre of Fusion Energy, UKAEA

https://github.com/qingfengxia
https://www.researchgate.net/profile/Qingfeng_Xia
personal website: https://www.iesensor.com

________________________________

**Comment by martinRenou** [comment]
Great! Thanks!

**Comment by qingfengxia** [comment]
For the cmake, I see pybind11 and json version is fixed, is there any special reason? I only manually test the code.  If testing on more versions is desired,  CI test matrix might be helpful.

**Comment by qingfengxia** [comment]
windows CI failed with "nlohmann/json.hpp" not found

**Comment by martinRenou** [comment]
Thanks a lot for updating the PR!

Let's just ignore the CI failure for now, I think I received emails from TravisCI saying that even master does not compile (but TravisCI does not report anything here anymore). I am going to remove appveyor and TravisCI, and use Github actions instead before making a release. So if I encounter this error I'll try and fix it.

Thanks for your PR.

**Comment by martinRenou** [comment]
@qingfengxia BTW, you are not getting any credit for your commits. If you want to be listed as contributors and get credits for your work on Github, you should add your `qingf*******@ukaea.uk` email address (that's the one specified in the commit) to your Github account. In the Github settings -> emails -> add email address

**Comment by martinRenou** [comment]
I can merge it right now. If you add your email address later it should still work. Thanks again for your PR!

**Comment by qingfengxia** [comment]
Thank you for the reminding, I prefer to keep email secret otherwise too much spam email.




Qingfeng Xia


https://github.com/qingfengxia
https://www.researchgate.net/profile/Qingfeng_Xia
personal website: https://www.iesensor.com

________________________________

**Pull request merged**